main.shares.tf

```diff
@@ -1,47 +1,39 @@
-resource "azurerm_storage_share" "this" {
+resource "azapi_resource" "share" {
   for_each = var.shares
 
-  name                 = each.value.name
-  quota                = each.value.quota
-  storage_account_name = azurerm_storage_account.this.name
-  access_tier          = each.value.access_tier
-  enabled_protocol     = each.value.enabled_protocol
-  metadata             = each.value.metadata
+  type = "Microsoft.Storage/storageAccounts/fileServices/shares@2023-01-01"
+  body = {
+    properties = {
+      metadata          = each.value.metadata
+      access_tier       = each.value.access_tier
+      enabledProtocols  = each.value.enabled_protocol
+      shareQuota        = each.value.quota
+      rootSquash        = each.value.rootSquash
```

Comment:
should be using snake casing on the inputs not camel casing (see similar comment on variables block).

e.g. 
rootSquash = each.value.root_squash